### PR TITLE
feat: Add config struct of OpenID-Connect Login

### PR DESCRIPTION
**Pull request by ljzsmllx**
Please answer these questions before submitting a pull request, **or your PR will get closed**.

**Why submit this pull request?**

- [x] Bugfix
- [x] New feature provided
- [ ] Improve performance
- [ ] Backport patches

**What changes will this PR take into?**

Please update this section with detailed description.

feat: The series PR will add a new feature to the Dashboard to make it support login through OpenID-Connect. 
This PR only adds some new config information to the config file.

fix: Fix some errors occurred in go-lint test

**Related issues**

**Checklist:**

- [x] Did you explain what problem does this PR solve? Or what new features have been added?
- [x] Have you added corresponding test cases?
- [x] Have you modified the corresponding document?
- [x] Is this PR backward compatible? If it is not backward compatible, please discuss on the mailing list first

### CODE REVIEW

File: api/conf/conf.yaml [hunk 1]
```diff
@@ -19,26 +19,26 @@
...
-                          # This value accepts IPv4, IPv6, and hostname.
+    # The default value is 0.0.0.0, if want to specify, please enable it.
+    # This value accepts IPv4, IPv6, and hostname.
     port: 9000            # The port on which the `Manager API` should listen.
 
-  # ssl:
-  #   host: 127.0.0.1     # the address on which the `Manager API` should listen for HTTPS.
-                          # The default value is 0.0.0.0, if want to specify, please enable it.
+      # ssl:
+      #   host: 127.0.0.1     # the address on which the `Manager API` should listen for HTTPS.
```
Comment (starsz) [inline api/conf/conf.yaml:27]: Why do you change this?

File: api/conf/conf.yaml [hunk 2]
```diff
@@ -71,15 +71,27 @@ conf:
...
+    # NOTE: Highly recommended to modify this value to protect `manager api`.
+  # if it's default value, when `manager api` start, it will generate a random string to replace it.
   expire_time: 3600     # jwt token expire time, in second
   users:                # yamllint enable rule:comments-indentation
     - username: admin   # username and password for login `manager api`
       password: admin
     - username: user
       password: user
 
+oidcApp:
```
Comment (starsz) [inline api/conf/conf.yaml:83]: `oidc` would be better.

File: api/conf/conf.yaml [hunk 3]
```diff
@@ -71,15 +71,27 @@ conf:
...
+  secret:
+    secret
+  expire_time: 3600
+  apps:
+    - appName: authing
+      clientId: 62f3bafc15fe957a20a2ab1a
+      clientSecret: 35e4d9c9a21d2176922a4b0395ec1373
```
Comment (starsz) [inline api/conf/conf.yaml:90]: Better not put your config here.

File: api/internal/conf/conf.go [hunk 4]
```diff
@@ -126,10 +128,26 @@ type Authentication struct {
 	Users      []User
 }
 
+type App struct {
+	AppName      string
+	ClientId     string
+	ClientSecret string
+	Scope        string
+	State        string
+	RedirectUri  string
```
Comment (starsz) [inline api/internal/conf/conf.go:137]: Should add `yaml` flag for each field.

File: api/conf/conf.yaml (hunk 1 above)
Comment (ljzsmllx) [reply api/conf/conf.yaml:27]: Sorry, hand shaking...

File: api/conf/conf.yaml (hunk 2 above)
Comment (ljzsmllx) [reply api/conf/conf.yaml:83]: OK

File: api/conf/conf.yaml [hunk 5]
```diff
@@ -51,14 +51,14 @@ conf:
...
+      # such as absolute path on Windows: winfile:///C:\error.log
     access_log:
       file_path:
         logs/access.log  # supports relative path, absolute path, standard output
-                         # such as: logs/access.log, /tmp/logs/access.log, /dev/stdout, /dev/stderr
-                         # such as absolute path on Windows: winfile:///C:\access.log
-                         # log example: 2020-12-09T16:38:09.039+0800	INFO	filter/logging.go:46	/apisix/admin/routes/r1	{"status": 401, "host": "127.0.0.1:9000", "query": "asdfsafd=adf&a=a", "requestId": "3d50ecb8-758c-46d1-af5b-cd9d1c820156", "latency": 0, "remoteIP": "127.0.0.1", "method": "PUT", "errs": []}
+        # such as: logs/access.log, /tmp/logs/access.log, /dev/stdout, /dev/stderr
+        # such as absolute path on Windows: winfile:///C:\access.log
+      # log example: 2020-12-09T16:38:09.039+0800	INFO	filter/logging.go:46	/apisix/admin/routes/r1	{"status": 401, "host": "127.0.0.1:9000", "query": "asdfsafd=adf&a=a", "requestId": "3d50ecb8-758c-46d1-af5b-cd9d1c820156", "latency": 0, "remoteIP": "127.0.0.1", "method": "PUT", "errs": []}
```
Comment (starsz) [inline api/conf/conf.yaml:61]: Better not modify this.

File: api/conf/conf.yaml [hunk 6]
```diff
@@ -71,15 +71,27 @@ conf:
 authentication:
   secret:
     secret              # secret for jwt token generation.
-                        # NOTE: Highly recommended to modify this value to protect `manager api`.
-                        # if it's default value, when `manager api` start, it will generate a random string to replace it.
+    # NOTE: Highly recommended to modify this value to protect `manager api`.
+  # if it's default value, when `manager api` start, it will generate a random string to replace it.
```
Comment (starsz) [inline api/conf/conf.yaml:75]: Better not modify this.

File: api/conf/conf.yaml [hunk 7]
```diff
@@ -71,15 +71,27 @@ conf:
...
+oidc:
+  secret:
+    secret
+  expire_time: 3600
+  apps:
+    - app_name: authing
+      client_id:
+      client_secret:
+      scope: oidc
+      state: 123456
```
Comment (starsz) [inline api/conf/conf.yaml:92]: What's the state field?

File: api/internal/conf/conf.go [hunk 8]
```diff
@@ -126,10 +128,26 @@ type Authentication struct {
 	Users      []User
 }
 
+type App struct {
+	AppName      string `mapstructure:"app_name"`
+	ClientId     string `mapstructure:"client_id"`
+	ClientSecret string `mapstructure:"client_secret"`
+	Scope        string
```
Comment (starsz) [inline api/internal/conf/conf.go:135]: Do we need to add `mapstructure` of this field?

File: api/conf/conf.yaml [hunk 9]
```diff
@@ -80,6 +80,17 @@ authentication:
     - username: user
       password: user
 
+oidc:
+  secret:
+    secret
+  expire_time: 3600
+  apps:
+    - app_name: authing
```
Comment (starsz) [inline api/conf/conf.yaml:88]: I think we can support one app first.
So just remove apps.

File: api/conf/conf.yaml [hunk 10]
```diff
@@ -80,6 +80,17 @@ authentication:
...
+      client_id:
+      client_secret:
+      scope: oidc
+      redirect_uri: /authing/callback
```
Comment (starsz) [inline api/conf/conf.yaml:92]: `redirect_uri` shouldn't contain the  `authing` config.
Maybe `oidc/callback`.

File: api/conf/conf.yaml (hunk 9 above)
Comment (ljzsmllx) [reply api/conf/conf.yaml:88]: ok

File: api/conf/conf.yaml (hunk 10 above)
Comment (ljzsmllx) [reply api/conf/conf.yaml:92]: ok

**Review by starsz — APPROVED** [review]

File: api/internal/handler/data_loader/route_import.go [hunk 11]
```diff
@@ -113,7 +113,6 @@ func (h *ImportHandler) Import(c droplet.Context) (interface{}, error) {
 			MergeMethod: input.MergeMethod == "true",
 			TaskName:    input.TaskName,
 		}
-		break
```
Comment (nic-chen) [inline api/internal/handler/data_loader/route_import.go:116]: why change this?

File: api/internal/conf/conf.go [hunk 12]
```diff
@@ -283,6 +298,14 @@ func initAuthentication(conf Authentication) {
 	}
 }
 
+func initOidc(conf Oidc) {
+	OidcConf = conf
+
+	if OidcConf.Secret == "secret" {
+		OidcConf.Secret = utils.GetFlakeUidStr()
```
Comment (nic-chen) [inline api/internal/conf/conf.go:305]: I think we should not set a random string to Secret. we could add a warning for user.

File: api/internal/handler/data_loader/route_import.go (hunk 11 above)
Comment (ljzsmllx) [reply api/internal/handler/data_loader/route_import.go:116]: The go-lint advised me to change this. I just followed it.

Comment (ljzsmllx) [reply api/internal/handler/data_loader/route_import.go:116]: I will change it back in next commit.

File: api/internal/conf/conf.go (hunk 12 above)
Comment (ljzsmllx) [reply api/internal/conf/conf.go:305]: After thinking twice, I think the `secret` is not necessary, I will delete it in the next commit.

**Review by nic-chen — APPROVED** [review]

**Comment by nic-chen** [comment]
Maybe we should merge to the `next` branch now? 
Looks like in later versions we will use `next` branch instead of `master`? 
@bzp2010 @starsz

**Comment by starsz** [comment]
> Maybe we should merge to the `next` branch now? Looks like in later versions we will use `next` branch instead of `master`? @bzp2010 @starsz

Since we will merge the `next` branch to `master`.
So both are OK.

**Pull request merged**